<a href="https://colab.research.google.com/github/denisparra/diplomadoInfoVis/blob/master/Diplomado_InfoVis_Alumnos_2019_Parte_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Diplomado _Big Data_ - Altair: Declarative Visualization in Python

**Profesores:** Nebil Kawas y Denis Parra.

**Ayudantes**: Juan Espinosa, Germán Contreras y Hernán Valdivieso.

Guía creada por Hernán Valdivieso.

Agradecimientos a Eduardo Graells (@carnby) por su codigo en github, sirvió como inspiración para varios ejercicios de datos con pandas.

-----

El siguiente _jupyter notebook_ tiene como fín ser una guía de cómo utilizar la librería de Python [Altair](https://altair-viz.github.io/index.html) y luego aplicar dicho conocimiento para responder una serie de dudas sobre un _dataset_ interesante: respuestas del [Censo 2017](http://www.censo2017.cl/microdatos/).

Para facilitar la instalación de esta librería, este archivo debe ser subido a [Google Colab](https://colab.research.google.com/) para poder ejecutarlo en los servidores de Google, los cuales ya disponen de todos los recursos necesarios para ejecutar los códigos que implementarás.


Este  _jupyter notebook_ se divide en 3 partes:

- Explorar los datos del _dataset_ entregado y aprender algunos usos de la librería [`pandas`](https://pandas.pydata.org/) para preprocesar los datos. 
- Entregar diversos ejemplos de algunos _idioms_, junto con una explicación detallada de que hace cada línea. Luego, para cada _idiom_ explicado, se solicitará que ustedes hagan otro gráfico con el _dataset_ del Censo y respondan una serie de preguntas. 

Puede realizar el práctico en parejas. Solo 1 de las parejas debe entregar.

### Formato de entrega: PDF

Para generar un pdf, ir a "Archivo", Imprimir.

Imprimir en orientación Vertical y a color.



### Nombres
- 
- 

## Indice

>[Diplomado Big Data - Altair: Declarative Visualization in Python](#scrollTo=r3VyfThGQdud)

>>[Indice](#scrollTo=z_YnPXBG1OYr)

>>[Descargar dataset](#scrollTo=9ckes_XsRkjN)

>>[Descripción del dataset](#scrollTo=fHHT_LhIAMBB)

>>[A explorar el dataset](#scrollTo=ops-zKxKWo7Q)

>>>[Actividad 1: exploración de dataset](#scrollTo=ops-zKxKWo7Q)

>>[Manipulacíón básica de los datos.](#scrollTo=7aAfiGfZWq80)

>>>[numpy y pandas.](#scrollTo=JwYdruuhWtQW)

>>>[Funciones de resumen](#scrollTo=4aJ9fXMkWuTE)

>>>>[Actividad 2: funciones de resumen](#scrollTo=Rq_MTug2XD5j)

>>>[Filtros](#scrollTo=dz4h0X8YhImo)

>>>>[Actividad 3: filtros](#scrollTo=O5cjZnLSXuDk)

>>>[Funciones de agregación](#scrollTo=a5oeGPlhhDHT)

>>>>[Actividad 4: funciones de agregación](#scrollTo=wbHP17qbYAF4)

>>>[Bonus: dask y joblib](#scrollTo=SSjFGsg2lJPy)

>>[Gráficos de ejemplo con Altair](#scrollTo=WAkmGyJPaImq)

>>>[Gráficos de barra](#scrollTo=srJQP-LJsHVM)

>>>>[Bar Chart](#scrollTo=Ekg96Eik3e9H)

>>>>>[Descripción del ejemplo](#scrollTo=MsyK29c9qPwI)

>>>>[Actividad 5: Bar chart](#scrollTo=0mRjsG9VqRAx)

>>>>[Grouped Bar Chart](#scrollTo=mVX2dOUdRcAv)

>>>>>[Descripción del ejemplo](#scrollTo=M3eOJ-9YA14a)

>>>>[Actividad 6: Grouped Bar Chart](#scrollTo=qrzyiUIPojBy)

>>>>[Stacked Bar Chart](#scrollTo=MBY1H13Czv60)

>>>>>[Descripción del ejemplo](#scrollTo=8RXnojZAqXJG)

>>>>[Actividad 7: Grouped Bar Chart](#scrollTo=J3tqoNKtpNkY)

>>>>[Series de tiempo](#scrollTo=Mz6VxzdtqlLI)

>>>>>[Descripción del ejemplo](#scrollTo=rTyJxIkaqlhK)

>>>>[Actividad 8: series de tiempo](#scrollTo=Z0E1x_xiqln1)



## Descargar _dataset_
En primero lugar, como estaremos ocupando un servidor externo, es necesario descargar el _dataset_ en dicho servidor. Para esto ocuparemos el comando `wget` el cual descagará desde dropbox el _dataset_ a utilizar.

In [0]:
!wget -N https://www.dropbox.com/s/0xp8srkxmfcfbb9/DatasetCenso.csv

--2019-05-04 17:37:01--  https://www.dropbox.com/s/0xp8srkxmfcfbb9/DatasetCenso.csv
Resolving www.dropbox.com (www.dropbox.com)... 162.125.8.1, 2620:100:601f:1::a27d:901
Connecting to www.dropbox.com (www.dropbox.com)|162.125.8.1|:443... connected.
HTTP request sent, awaiting response... 301 Moved Permanently
Location: /s/raw/0xp8srkxmfcfbb9/DatasetCenso.csv [following]
--2019-05-04 17:37:01--  https://www.dropbox.com/s/raw/0xp8srkxmfcfbb9/DatasetCenso.csv
Reusing existing connection to www.dropbox.com:443.
HTTP request sent, awaiting response... 302 Found
Location: https://uc1192a04b120f4f9b17885e2ae3.dl.dropboxusercontent.com/cd/0/inline/AgQSIKa45kX96agfcCeNnrPw-qdfNweZmORYH2EsV6P4t16vN0LywwO435Fyb4ZmwLInWz-fkTaWfsUfYaCDHl6fDWmCUGApWAE7pCbBwHcTuw/file# [following]
--2019-05-04 17:37:01--  https://uc1192a04b120f4f9b17885e2ae3.dl.dropboxusercontent.com/cd/0/inline/AgQSIKa45kX96agfcCeNnrPw-qdfNweZmORYH2EsV6P4t16vN0LywwO435Fyb4ZmwLInWz-fkTaWfsUfYaCDHl6fDWmCUGApWAE7pCbBwHcTuw/file
Resolvi

El _dataset_ ha sido descargado, si ahora utilizamos el comando `ls` podremos observar que archivos están actualmente en el servidor y que podrás utilizar desde este _jupyter notebook_.

In [0]:
ls

DatasetCenso.csv  sample_data/


## Descripción del _dataset_
Con el fín de reducir el peso de este archivo, fue necesario eliminar multiples columnas del _dataset_ original. Por lo tanto, las columnas de las preguntas conservadas y su respectiva descripción son:

- **comuna**: comuna de la casa que fue censada.
- **parentesco**:  relación de parentesco con el jefe/a de hogar. Esta columna corresponde a la columna P07 del _dataset_ original.
- **edad**: años cumplidos. Esta columna corresponde a la columna P09 del _dataset_ original.
- **pais**: país donde vivía la madre del encuestado cuando nació. Esta columna corresponde a la columna P12PAIS del _dataset_ original.
- **a_llegada**:  año que llegó el encuestado a Chile. Esta columna corresponde a la columna P12A_LLEGADA del _dataset_ original.
- **n_educacional**:  nivel educacional logrado del encuestado. Esta columna corresponde a la columna P15 del _dataset_ original luego de pasar por una función que agrupó diferentes niveles educaciones a un número fijo.
- **descendencia**: cantidad de hijos del encuestado. Esta columna corresponde a la columna P19 del _dataset_ original.


**Importante**: Las columnas están separadas por punto y coma (`";"`).

Adicionalmente, para reducir la cantidad de respuestas, se realizó un filtro para dejar solo las respuestas dadas por personas inmigrantes.




## A explorar el _dataset_

### Actividad 1: exploración de _dataset_

Antes de empezar a procesar y visualizar los datos, es necesario entender el _dataset_. Por lo tanto, primero necesitamos que identifiques las columnas que son **cuantitativas**, **categóricas** y **ordinales**.


** Respuesta **

COMPLETAR





## Manipulacíón básica de los datos.

Para poder describir el _dataset_ se introducirá el uso de la librería [`pandas`](https://pandas.pydata.org/), el cual es un paquete de Python que provee de una serie de funciones, métodos y estructuras de datos para trabajar los _dataset_ de forma facil e intuitiva.



### `numpy` y `pandas`.
Primero, es necesario saber existe una librería para Python llamada [`numpy`](http://www.numpy.org/) la cual  permite crear arreglos y matrices multidimensionales que contienen datos genéricos. Dichos arreglos y matrices son operados de forma eficiente en memoria e incrementa en gran medida la velocidad de las operaciones realizadas con cada uno de sus elementos. `pandas` está construido sobre la librería de `numpy`. Esto produce que algunos métodos o funciones de `pandas` no retornen tipos de datos creados en la propia librería, como los [`DataFrame`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.html) o [`Series`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.html), sino que retorne estructuras de ``numpy`` como el  [``numpy.ndarray``](https://docs.scipy.org/doc/numpy-1.14.0/reference/generated/numpy.ndarray.html). Al mismo tiempo, ``pandas`` puede cargar facilmente matrices que provengan de `numpy`.





### Funciones de resumen

Lo primero que necesitamos es poder cargar el _dataset_ con `pandas` y luego lograr obtener alguna información sobre sus columnas. Para esto se enseñará cómo abrir y obtener algún dato mínimo para luego solicitar que hagas acciones similares con otras columnas.

- **Abrir un archivo CSV** Para realizar tal acción se utiliza la función [`read_csv`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.read_csv.html) que retorna un `DataFrame` con la información del archivo. A continuación vamos a cargar el archivo y guardarlo en la variable `censo`.





In [0]:
import pandas as pd
censo = pd.read_csv("DatasetCenso.csv", sep=";")

- **Obtener información de resumen:** a veces es necesario explorar los datos, como saber cuales son los valores posibles de alguna columna categórica o saber el valor mínimo de una columna numérica. A continuación vamos a ver cuál es el año mínimo en el que han llegado los inmigrantes y de que cuántos países lo han hecho. Para lo primero, se selecionará la columna `a_llegada` ejecutando `censo.a_llegada` y luego se aplicará el método [`min`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.min.html) que retorna el valor mínimo de la columna seleccionada.

In [0]:
censo.a_llegada.min()

1950

Aquí podemos notar que el año en donde llegó el primer inmigrante según este censo es en 1950. Ahora, para obtener la cantidad de paises de los cualen han llegado inmigrantes, es necesario utilizar más de un línea de código. Primero vamos a encontrar los valores únicos de la columna `pais` y luego contar la cantidad de países. El código para obtener tal información es:

In [0]:
paises = censo.pais.unique()
paises_dataframe = pd.DataFrame(paises)
paises_dataframe.shape

(76, 1)

La respuesta es que existen 76 paises distintos. Para obtener este resultado se realizaron 3 líneas de código:
1. `paises = censo.pais.unique()`:   en esta línea se realizan 2 acciones. Primero se le indica al `DataFrame` `censo` que seleccione la columna `pais`  mediante el código `censo.pais` y a dicha comuna se le aplica el método [`unique`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.unique.html) el cual retorna un `numpy.ndarray` con todos los valores diferentes de esa columna.
2. `paises_dataframe = pd.DataFrame(paises)`: con esta línea, se está transformando el `numpy.ndarray` en un `DataFrame`, esto último se hace porque `pandas` posee una variedad de métodos mucho mayor para el procesamiento de datos que `numpy`.
3. `paises_dataframe.shape`:  con el atributo [`shape`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.shape.html), se obtiene la dimensión de la matriz. Como `unique` nos entregó una lista de varios elementos, la dimensión es solo 1 fila con múltiples elementos. Eso explica el `(76, 1)`, el primero es la cantidad de columnas distintas y el segundo es la cantidad de filas, es este caso, 1.



#### Actividad 2: funciones de resumen


Ahora, utilizando la librería ``pandas`` necesitamos que escriba el código necesario para realizar lo siguiente:

1. Responder: ¿En cuántas comunas distintas fueron censados los inmigrantes? 
2. Indicar cuáles son las 10 primeras y 10 últimas entrevistas registradas en el censo.
3. Describir cada columna numérica y ordinal con: la cantidad de datos, promedio, desviación estandar, mínimo, máximo y los diferentes cuartiles.

**Respuesta pregunta 1**

COMPLETAR




**Respuesta pregunta 2**

COMPLETAR

**Hint:** investigue el uso de los métodos [`head`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.head.html) y [`tail`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.tail.html).

**Respuesta pregunta 3**

COMPLETAR

**Hint:** investigue el uso de `pandas` y del método [`describe`]().

**¿Hay algo raro con la descripción?** ¿Puede ser que algún dato no tenga mucho sentido? Esto pasa porque dentro del _dataset_ hay valores que no representan la respuesta del encuestado, sino que indica que dicha respuesta no aplica a lo preguntado o no se obtuvo una respuesta, esto se codificó como 99 o 98  en las columnas  `n_educacional` y `descendencia`, y como 9999 en `a_llegada`. Todos estos datos siempre se exigen filtrar para realizar una visualización que no sea afectada por el ruido de personas con datos fuera de lo esperado.

### Filtros
En muchos casos, se require filtrar por uno o más datos. A continuación se mostrará un ejemplo de como filtrar para obtener la cantidad de inmigrantes de un solo país y luego de una lista de paises. En particular, queremos saber la cantidad de inmigrantes de México y de Japón o Alemania

In [0]:
censo[censo.pais == "México"].count()

comuna           2843
parentesco       2843
edad             2843
pais             2843
a_llegada        2843
n_educacional    2843
descendencia     2843
dtype: int64

En este caso,  con el código `censo[censo.pais == "México"]`, le estamos indicando al DataFrame que solo retorne aquellas filas donde la columa `pais` sea igual a `"México"`. Posteriormente estamos aplicando el método [`count`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.count.html) para saber cuantos resultados persisten por columna. Para más información de como seleccionar información, pueden revisar esta [página](https://pandas.pydata.org/pandas-docs/stable/indexing.html).

Ahora, para obtener la cantidad de inmigrantes de una lista de paises, hay dos formas, realizar una busqueda similar a la anterior, pero agregando `or` o utilizar el método [`isin`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.isin.html). En este caso ocuparemos la segunda opción:


In [0]:
censo[censo.pais.isin(["Japón", "Alemania"])].count()

comuna           3538
parentesco       3538
edad             3538
pais             3538
a_llegada        3538
n_educacional    3538
descendencia     3538
dtype: int64

En este caso, disponemos de 3538 personas cuyo origin es Japón o Alemania. 



#### Actividad 3: filtros
1. Indique la cantidad de inmigrantes que llegaron antes del 2010.
2. Indique la cantidad de inmigrantes que no fueron censados en Macul o Puente Alto.


**Respuesta pregunta 1**

COMPLETAR

**Hint:** investigue el uso del simbolo "<" en ``pandas`` para filtrar.


**Respuesta pregunta 2**

COMPLETAR

**Hint:** investigue en google cómo se aplica la negación en los filtros en `pandas`.


### Funciones de agregación

Finalmente, un uso práctico de ``pandas`` es agrupar columnas bajo 1 o más parametros en común. Para esto se usa la función [`groupby`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.groupby.html) que recibe una lista de las columnas y genera grupos donde todas las filas de cada uno comparten con la misma combinación de valores en las comunas indicadas. Por ejemplo, vamos a contar la cantidad de inmigrantes por nivel educacional.


**Nota adicional**: Más información de `groupby` [aquí](https://pandas.pydata.org/pandas-docs/stable/groupby.html)

In [0]:
censo.groupby("n_educacional").count()

comuna  parentesco   edad   pais  a_llegada  descendencia
n_educacional                                                           
1                 116         116    116    116        116           116
3                9843        9843   9843   9843       9843          9843
6               34044       34044  34044  34044      34044         34044
9               57798       57798  57798  57798      57798         57798
12              10297       10297  10297  10297      10297         10297
98               3928        3928   3928   3928       3928          3928

Aquí podemos observar que la mayor cantidad de personas obtiene un nivel educacional de 9 seguido del 6, es decir, la mayoría de los inmigrantes tiene educación media o superior finalizada. 


#### Actividad 4: funciones de agregación

Ahora, utilice esta función y lo aprendido anteriormente para indicar la edad promedio de los inmigrantes por país y nivel educacional de únicamente los jefes de hogar. 


**Respuesta**

COMPLETAR

**Hint:** Aquí es necesario **filtrar**, luego **agrupar** por más valores y utilizar otro método similar a `count` para **calcular el promedio del grupo**.

### Bonus: [`dask`](http://dask.pydata.org/en/latest/docs.html) y [`joblib`](https://pythonhosted.org/joblib/index.html)

Anteriormente mostramos métodos para el filtro y agregación de datos, pero ¿Qué pasa si el _dataset_ pesa 4GB o más? ¿Se podrá filtrar de forma eficiente?.  ¿Sigo usando pandas?

Para estos casos, no se recomienda utilizar pandas porque este intentará dejar en memoria RAM todo el _dataset_ cargado, y podria ocurrir que no tienes suficiente memoria en tu computador. Es por ello que te contamos de 2 librerías de Python que buscan entregar una forma rápida  e intuitiva para trabajar con grandes volumenes de datos. En particular, dask utiliza los mismos métodos y funciones de pandas, pero por detrás posee un sistema capaz de, entre otras cosas: a) procesar archivos que no caben en memoria, y b) paralelizar todas las acciones que deseas realizar (filtrar, seleccionar, agrupar, etc) en diferentes _cores_ de tus CPU. Si bien no es el alcance de este práctico, tenlos en cuenta cuando quieras procesar datos en el futuro.

## Gráficos de ejemplo con Altair

### Gráficos de barra


#### [Bar Chart](https://altair-viz.github.io/gallery/simple_bar_chart.html)

In [0]:
import altair as alt
import pandas as pd

# alt.renderers.enable('notebook')


data = pd.DataFrame({
    'a': ['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I'],
    'b': [55, 48, 34, 21, 11, 73, 89, 97, 100]
})

alt.Chart(data).mark_bar().encode(
    x='a',
    y='b'
)

Chart({
  data:    a    b
  0  A   55
  1  B   48
  2  C   34
  3  D   21
  4  E   11
  5  F   73
  6  G   89
  7  H   97
  8  I  100,
  encoding: FacetedEncoding({
    x: X({
      shorthand: 'a'
    }),
    y: Y({
      shorthand: 'b'
    })
  }),
  mark: 'bar'
})

##### Descripción del ejemplo

Ahora procederemos a explicar línea a línea el código. 
-  **Línea 1 y 2**

```
import altair as alt
import pandas as pd
```
Con estas 2 líneas estamos importanto la librería ``altair`` y la librería ``pandas``. Es de vitar importantia partir importando todo módulo o librería externa a Python que utilizarás para no generar error.

- **Línea 4 a 7**

```python
data = pd.DataFrame({
    'a': ['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I'],
    'b': [55, 48, 34, 21, 11, 73, 89, 97, 100]
})
```
Con estas línea, se está creando un DataFrame de 2 columnas (`a` y `b`) en donde `a` serán datos categoricos y `b` serán datos numéricos.
- **Línea 9**

```python
alt.Chart(data).mark_bar().encode(
```
En esta línea se hacen 2 acciones:
1. Se entrega el  ``DataFrame``  `data` al objeto [`chart`](https://altair-viz.github.io/user_guide/API.html#altair.Chart). Así se le indica a la visualización de donde se obtendrán los datos.
2. Se setea el tipo de marca para representar el _dataset_ con el método `mark_bar`. Este método indica que se representará mediante barras. Si quieres ver diferentes marcas, en esta [página](https://altair-viz.github.io/user_guide/marks.html) puedes encontrar más.
    1. Todos los métodos de marcas pueden recibir argumentos para personalizar dichas marcas, uno que veremos más adelante es el _stroke_ para editar el borde de las barras.
    
Adicionalmente, se agrega el `.encode(` para indicar que ahora se comenzará a entregar líneas de código de como será el _encoding_ de los datos en el gráfico.

- **Línea 10 y 11**

```python
    x='a',
    y='b',
```
Con estas 2 líneas, se indica el _encoding_ de los ejes, en particular le estamos indicando las columna del `DataFrame` a utilizar para cada eje.

#### Actividad 5: _Bar chart_
Seleccione 5 países e indique la cantidad de inmigrantes por pais mediante un gráfico de barras contruido con Altair. 

**Importante** Tenga precaución del número máximo de filas soportadas por Altair.

#### [Grouped Bar Chart](https://altair-viz.github.io/gallery/grouped_bar_chart.html)

In [0]:
import altair as alt
from altair.expr import datum, if_
from vega_datasets import data

source = data.population.url
data = pd.read_json(source)

alt.Chart(data).mark_bar(stroke='blue').encode(
    alt.X('gender:N', scale=alt.Scale(rangeStep=8), axis=alt.Axis(title='')),
    alt.Y('sum(people):Q', axis=alt.Axis(title='population', grid=True)),
    color=alt.Color('gender:N', scale=alt.Scale(range=["#EA98D2", "#659CCA"])),
    column='age:O'
).configure_view(stroke='transparent'
).configure_axis(domainWidth=0.8
).transform_filter(datum.year == 2000
).transform_calculate('gender', if_(datum.sex == 2, 'Female', 'Male'))

Chart({
  config: Config({
    axis: AxisConfig({
      domainWidth: 0.8
    }),
    view: ViewConfig({
      stroke: 'transparent'
    })
  }),
  data:      age    people  sex  year
  0      0   1483789    1  1850
  1      0   1450376    2  1850
  2      5   1411067    1  1850
  3      5   1359668    2  1850
  4     10   1260099    1  1850
  5     10   1216114    2  1850
  6     15   1077133    1  1850
  7     15   1110619    2  1850
  8     20   1017281    1  1850
  9     20   1003841    2  1850
  10    25    862547    1  1850
  11    25    799482    2  1850
  12    30    730638    1  1850
  13    30    639636    2  1850
  14    35    588487    1  1850
  15    35    505012    2  1850
  16    40    475911    1  1850
  17    40    428185    2  1850
  18    45    384211    1  1850
  19    45    341254    2  1850
  20    50    321343    1  1850
  21    50    286580    2  1850
  22    55    194080    1  1850
  23    55    187208    2  1850
  24    60    174976    1  1850
  25    60    162236    2  1850
  26    65    106827    1  1850
  27    65    105534    2  1850
  28    70     73677    1  1850
  29    70     71762    2  1850
  ..   ...       ...  ...   ...
  540   20   9731315    1  2000
  541   20   9324244    2  2000
  542   25   9659493    1  2000
  543   25   9518507    2  2000
  544   30  10205879    1  2000
  545   30  10119296    2  2000
  546   35  11475182    1  2000
  547   35  11635647    2  2000
  548   40  11320252    1  2000
  549   40  11488578    2  2000
  550   45   9925006    1  2000
  551   45  10261253    2  2000
  552   50   8507934    1  2000
  553   50   8911133    2  2000
  554   55   6459082    1  2000
  555   55   6921268    2  2000
  556   60   5123399    1  2000
  557   60   5668961    2  2000
  558   65   4453623    1  2000
  559   65   4804784    2  2000
  560   70   3792145    1  2000
  561   70   5184855    2  2000
  562   75   2912655    1  2000
  563   75   4355644    2  2000
  564   80   1902638    1  2000
  565   80   3221898    2  2000
  566   85    970357    1  2000
  567   85   1981156    2  2000
  568   90    336303    1  2000
  569   90   1064581    2  2000
  
  [570 rows x 4 columns],
  encoding: FacetedEncoding({
    color: Color({
      scale: Scale({
        range: ['#EA98D2', '#659CCA']
      }),
      shorthand: 'gender:N'
    }),
    column: Column({
      shorthand: 'age:O'
    }),
    x: X({
      axis: Axis({
        title: ''
      }),
      scale: Scale({
        rangeStep: 8
      }),
      shorthand: 'gender:N'
    }),
    y: Y({
      axis: Axis({
        grid: True,
        title: 'population'
      }),
      shorthand: 'sum(people):Q'
    })
  }),
  mark: MarkDef({
    stroke: 'blue',
    type: 'bar'
  }),
  transform: [FilterTransform({
    filter: (datum.year === 2000)
  }), CalculateTransform({
    as: 'gender',
    calculate: if((datum.sex === 2),'Female','Male')
  })]
})

##### Descripción del ejemplo

Ahora procederemos a explicar línea a línea el código, si alguna ya fue explicada en un gráfico anterior, se omitirá su explicación
-  **Línea 2 y 3**

```
from altair.expr import datum, if_
from vega_datasets import data
```
Con estas 2 líneas estamos importanto algunas funciones de altair (`datum` e `if_`) y el módulo `data` el cual contiene diferentes _datasets_ para poder utilizar. 

----------------------------
- **Línea 4 y 5**

```python
source = data.population.url
data = pd.read_json(source)
```
Con esta línea, se está guardando en la variable `source` el  _link_ a un archivo [JSON](https://www.json.org/). Precisamente estamos guardando el link al [siguiente archivo](https://vega.github.io/vega-datasets/data/population.json). Como podrás ver, es una lista de diccionario donde cada diccionario sigue el siguiente formato:

``` {"year":número,"age":número,"sex":número (1 o 2),"people":número}, ```

Con `data = pd.read_json(source)` estamos creando un `DataFrame` de pandas con el contenido de dicho _link_.

----------------
- **Línea 7**

```python
alt.Chart(data).mark_bar(stroke='blue').encode(
```
En esta línea se hacen 2 acciones ya vistas (entregar el _dataset_ y setear el tipo de marca), pero se agrega el uso de argumentos dentro del método `mark_bar`. En este caso,  se está seteando una propiedad a la marca con `stroke='blue'`. Esto hace que el borde de las barras sean de un color en especifico, en particular, azul.

----------------


- **Línea 8**

```python
alt.X('gender:N', scale=alt.Scale(rangeStep=8), axis=alt.Axis(title='')),
```
En esta línea, se describe el _encoding_ del eje X, como queremos indicar más propiedades de dicho eje, ya no basta con hacer `x='gender'` como el caso anterior, sino que utilizar [`alt.X`](https://altair-viz.github.io/user_guide/API.html?#altair.X) para indicar más propiedades.

Para este ejemplo, se indica que el eje X será la columna `gender` que es un dato nominal, eso justifica el `":N"`. Puedes revisar los otros típos de datos en el siguiente [link](https://altair-viz.github.io/user_guide/encoding.html#data-types). Se puede apreciar que `gender` no exista en el _dataset_ cargado, pero esto será explicado con la última línea del código que se encarga de crear dicha columna.

Junto a indicar el _encoding_, tambien se setean 2 propiedades de dicho eje.
- **`scale`**: este argumento permite indicar la escala a utilizar en dicho eje, en específico, con el `rangeStep` le estamos indicando el rango de espacio que ocupará cada barra.
- **`axis`**: este argumento recibe un objeto [`Axis`](https://altair-viz.github.io/user_guide/API.html?#altair.Axis) que permite describir dicho eje, en este caso, este objeto solo está describiendo que el título del eje sea vacío.
-------------
- **Línea 9**

```python
alt.Y('sum(people):Q', axis=alt.Axis(title='population', grid=True)),
```
    
En esta línea, se describe el _encoding_ del eje Y. En este caso se dice que será la sumatoria de la columna `people` el cual es un dato cuantitativo, eso justifica el `":Q"`. Es necesario destacar que estamos aplicando sumatoria, esto implica que el código en si está agrupando los datos y a cada grupo obtenido, le obtiene la sumatoria de dicha columna. Para este ejemplo estamos agrupando por género y por edad.

Junto a indicar el _encoding_, tambien se setea propiedad de `axis`, que le está indicando al eje que su título será `'population'` y que existan grillas, es decir, que aparezcan líneas de fondo para poder ubicar más rápido el valor de alguna barra.

-------------
- **Línea 10**

```python
color=alt.Color('gender:N', scale=alt.Scale(range=["#EA98D2", "#659CCA"])),
```

Esta línea indica que el color de las barras será representado bajo el objeto [`alt.Color`](https://altair-viz.github.io/user_guide/API.html#altair.Color) cuyo color  será definido según el valor  de la columna _gender_ y de _scale_ le indicamos que el rango de colores serán los 2 proporcionados ("#EA98D2" y "#659CCA")

-------------
- **Línea 11**

```python
column='age:O'
```
Esta última columna dentro de _encode_ indica que el parametro _age_ será el utilizado para separar cada grupo de columnas. En caso de omitir este dato, estaremos generando un gráfico de barra simple porque no estamos separando en diferentes grupos.

-------------
- **Línea 12**

```python
).configure_view(stroke='transparent'
```
El método [`configure_view`](https://altair-viz.github.io/user_guide/configuration.html#config-view) está personalizando la vista del gráfico, en este ejemplo le indica que los bordes de cada grupo de columnas sea transparente. Se recomienda cambiar el parametro a otro color (como _red_) para apreciar el efecto de dicho método.

-------------
- **Línea 13**

```python
).configure_axis(domainWidth=0.8
```

El método [`configure_axis`](https://altair-viz.github.io/user_guide/configuration.html#config-axis) está personalizando la vista de los ejes, en este caso le indica que tamaño de las líneas de los ejes sea de 0.8. Se recomienda cambiar el parametro a otro tamaño (como 10) para apreciar el efecto de dicho método.

-------------
- **Línea 14**

```python
).transform_filter(datum.year == 2000
```
Una de las acciones que usualmente se requiere en visualización, que es la [transformación de la información](https://altair-viz.github.io/user_guide/transform.html). En este caso, filtrar por uno o más valores.

El método [`transform_filter`](https://altair-viz.github.io/user_guide/transform.html#filter-transform) permte que la visualizacion filtre por dichos valores. Para esta ocasión utilizamos el objeto `datum` que fue importado para darselo a `transform_filter` e indicar que filtre por la columna _year_ cuyo valor debe ser igual a 2000.

En caso de querer filtrar por más de un atributo, se recomienda estudiar los [`Field Predicates
`](https://altair-viz.github.io/user_guide/transform.html#field-predicates)
                               

-------------
- **Línea 15**

```python
).transform_calculate('gender', if_(datum.sex == 2, 'Female', 'Male'))
```

`transform_calculate` es otro de los métodos dentro de la transformación de información. Este permite crear una nueva columna en base a una operación realizada a otras columnas ya existentes. Para este ejemplo, estamos creando la columna _gender_ que tomará 2 valores: _Female_ en caso de que _sex_ sea igual a 2 o _Male_ en otro caso.

---------
Podrás notar que en este ejemplo, no solo nos preocupamos de visualizar los datos, sino que de personalizar dicho gráfico con tamaños de línea, agregar grillas, color de los bordes, etc. Esto es muy importante para la percepción porque en caso de tener muchos datos,  los canales a usar y como presentarlos en el gráfico tendrán harta influencia en lo que el receptor entienda de tus datos.

#### Actividad 6: _Grouped Bar Chart_

¿Cual es la edad promedio de la gente que proviene de Japón, Haití, Cuba, África o Brasil y que fueron censados en Valdivia, Viña del mar o Puente Alto? Recuerde mostrar para cada combinación de país y comuna la edad promedio.


#### [Stacked Bar Chart](https://altair-viz.github.io/gallery/stacked_bar_chart.html)

In [0]:
import altair as alt
from vega_datasets import data

weather = data.seattle_weather()

alt.Chart(weather).mark_bar().encode(
    alt.Color('weather:N',
        legend=alt.Legend(title='Weather type'),
        scale=alt.Scale(
            domain=['sun', 'fog', 'drizzle', 'rain', 'snow'],
            range=['#e7ba42', '#c7c7c7', '#aec7e8', '#1f77b4', '#9467bd']
        ),
    ),
    alt.X('month(date):N', axis=alt.Axis(title='Month of the Year')),
    y='count()',
)

Chart({
  data:            date  precipitation  temp_max  temp_min  wind  weather
  0    2012-01-01            0.0      12.8       5.0   4.7  drizzle
  1    2012-01-02           10.9      10.6       2.8   4.5     rain
  2    2012-01-03            0.8      11.7       7.2   2.3     rain
  3    2012-01-04           20.3      12.2       5.6   4.7     rain
  4    2012-01-05            1.3       8.9       2.8   6.1     rain
  5    2012-01-06            2.5       4.4       2.2   2.2     rain
  6    2012-01-07            0.0       7.2       2.8   2.3     rain
  7    2012-01-08            0.0      10.0       2.8   2.0      sun
  8    2012-01-09            4.3       9.4       5.0   3.4     rain
  9    2012-01-10            1.0       6.1       0.6   3.4     rain
  10   2012-01-11            0.0       6.1      -1.1   5.1      sun
  11   2012-01-12            0.0       6.1      -1.7   1.9      sun
  12   2012-01-13            0.0       5.0      -2.8   1.3      sun
  13   2012-01-14            4.1       4.4       0.6   5.3     snow
  14   2012-01-15            5.3       1.1      -3.3   3.2     snow
  15   2012-01-16            2.5       1.7      -2.8   5.0     snow
  16   2012-01-17            8.1       3.3       0.0   5.6     snow
  17   2012-01-18           19.8       0.0      -2.8   5.0     snow
  18   2012-01-19           15.2      -1.1      -2.8   1.6     snow
  19   2012-01-20           13.5       7.2      -1.1   2.3     snow
  20   2012-01-21            3.0       8.3       3.3   8.2     rain
  21   2012-01-22            6.1       6.7       2.2   4.8     rain
  22   2012-01-23            0.0       8.3       1.1   3.6     rain
  23   2012-01-24            8.6      10.0       2.2   5.1     rain
  24   2012-01-25            8.1       8.9       4.4   5.4     rain
  25   2012-01-26            4.8       8.9       1.1   4.8     rain
  26   2012-01-27            0.0       6.7      -2.2   1.4  drizzle
  27   2012-01-28            0.0       6.7       0.6   2.2     rain
  28   2012-01-29           27.7       9.4       3.9   4.5     rain
  29   2012-01-30            3.6       8.3       6.1   5.1     rain
  ...         ...            ...       ...       ...   ...      ...
  1431 2015-12-02            2.5      10.6       4.4   5.0      fog
  1432 2015-12-03           12.7      15.6       7.8   5.9      fog
  1433 2015-12-04            2.0      10.6       6.1   4.7      fog
  1434 2015-12-05           15.7      10.0       6.1   4.0      fog
  1435 2015-12-06           11.2      12.8       7.2   5.9      fog
  1436 2015-12-07           27.4      11.1       8.3   3.4      fog
  1437 2015-12-08           54.1      15.6      10.0   6.2      fog
  1438 2015-12-09           13.5      12.2       7.8   6.3      fog
  1439 2015-12-10            9.4      11.7       6.1   7.5      fog
  1440 2015-12-11            0.3       9.4       4.4   2.8      sun
  1441 2015-12-12           16.0       8.9       5.6   5.6      fog
  1442 2015-12-13            1.3       7.8       6.1   6.1      sun
  1443 2015-12-14            0.0       7.8       1.7   1.7      sun
  1444 2015-12-15            1.5       6.7       1.1   2.9      fog
  1445 2015-12-16            3.6       6.1       2.8   2.3      fog
  1446 2015-12-17           21.8       6.7       3.9   6.0      fog
  1447 2015-12-18           18.5       8.9       4.4   5.1      fog
  1448 2015-12-19            0.0       8.3       2.8   4.1      fog
  1449 2015-12-20            4.3       7.8       4.4   6.7      fog
  1450 2015-12-21           27.4       5.6       2.8   4.3      fog
  1451 2015-12-22            4.6       7.8       2.8   5.0      fog
  1452 2015-12-23            6.1       5.0       2.8   7.6      fog
  1453 2015-12-24            2.5       5.6       2.2   4.3      fog
  1454 2015-12-25            5.8       5.0       2.2   1.5      fog
  1455 2015-12-26            0.0       4.4       0.0   2.5      sun
  1456 2015-12-27            8.6       4.4       1.7   2.9      fog
  1457 2015-12-28            1.5       5.0

##### Descripción del ejemplo

Ahora procederemos a explicar línea a línea el código, si alguna ya fue explicada en un gráfico 

- **Línea 4**

```python
weather = data.seattle_weather()
```
Con esta línea, se está guardando en la variable `weather` un DataFrame con datos del clima.

----------------
- **Línea 6 al 13**

```python
alt.Color('weather:N',
        legend=alt.Legend(title='Weather type'),
        scale=alt.Scale(
            domain=['sun', 'fog', 'drizzle', 'rain', 'snow'],
            range=['#e7ba42', '#c7c7c7', '#aec7e8', '#1f77b4', '#9467bd']
        ),
    ),```
En esta líneas se setea el color del gráfico.
1. `weather:N'` : aquí se indica que el color será definido según el valor de la columna_weather_ que es un dato nominal.
2. legend=alt.Legend(title='Weather type'): aquí se utiliza un objeto [`legend`](https://altair-viz.github.io/user_guide/API.html#altair.Legend) para definir el título de la leyenda del gráfico.
3. scale=alt.Scale(...): aquí se está seteando la escala de colores, en particular, se indica cuales son los valores que serán el dominio (`['sun', 'fog', 'drizzle', 'rain', 'snow']`) y para cada uno, con que color será representado.

En caso de omitir el objeto `Color`, estaríamos construyendo un gráfico de barra simple.

----------------


- **Línea 14 y 15**

```python
alt.X('month(date):N', axis=alt.Axis(title='Month of the Year')),
y='count()',

```
Estas lineas están seteando cada eje. En la primera se indica que tome el mes de cada fecha (`month(date)`) y que el nombre de ese eje sea _Month of the Year_. En la segunda línea se utiliza una [función de agregación (`count`)](https://altair-viz.github.io/user_guide/encoding.html?#encoding-channel-options) para indicar que tome la cantidad de filas que coindicen con el mes.

#### Actividad 7: _Grouped Bar Chart_
Para los inmigrantes de Japón, Haití, Cuba, África o Brasil que fueron censados en: Valdivia, Viña del mar o Puente Alto, indique el total de hijos que tiene cada padre. Utilice la columna _descendencia_.

#### Series de tiempo

In [0]:
import altair as alt
from vega_datasets import data

stocks = data.stocks()

alt.Chart(stocks).mark_line().encode(
    x='date',
    y='price',
    color='symbol'
).interactive()

Chart({
  data:     symbol       date   price
  0     MSFT 2000-01-01   39.81
  1     MSFT 2000-02-01   36.35
  2     MSFT 2000-03-01   43.22
  3     MSFT 2000-04-01   28.37
  4     MSFT 2000-05-01   25.45
  5     MSFT 2000-06-01   32.54
  6     MSFT 2000-07-01   28.40
  7     MSFT 2000-08-01   28.40
  8     MSFT 2000-09-01   24.53
  9     MSFT 2000-10-01   28.02
  10    MSFT 2000-11-01   23.34
  11    MSFT 2000-12-01   17.65
  12    MSFT 2001-01-01   24.84
  13    MSFT 2001-02-01   24.00
  14    MSFT 2001-03-01   22.25
  15    MSFT 2001-04-01   27.56
  16    MSFT 2001-05-01   28.14
  17    MSFT 2001-06-01   29.70
  18    MSFT 2001-07-01   26.93
  19    MSFT 2001-08-01   23.21
  20    MSFT 2001-09-01   20.82
  21    MSFT 2001-10-01   23.65
  22    MSFT 2001-11-01   26.12
  23    MSFT 2001-12-01   26.95
  24    MSFT 2002-01-01   25.92
  25    MSFT 2002-02-01   23.73
  26    MSFT 2002-03-01   24.53
  27    MSFT 2002-04-01   21.26
  28    MSFT 2002-05-01   20.71
  29    MSFT 2002-06-01   22.25
  ..     ...        ...     ...
  530   AAPL 2007-10-01  189.95
  531   AAPL 2007-11-01  182.22
  532   AAPL 2007-12-01  198.08
  533   AAPL 2008-01-01  135.36
  534   AAPL 2008-02-01  125.02
  535   AAPL 2008-03-01  143.50
  536   AAPL 2008-04-01  173.95
  537   AAPL 2008-05-01  188.75
  538   AAPL 2008-06-01  167.44
  539   AAPL 2008-07-01  158.95
  540   AAPL 2008-08-01  169.53
  541   AAPL 2008-09-01  113.66
  542   AAPL 2008-10-01  107.59
  543   AAPL 2008-11-01   92.67
  544   AAPL 2008-12-01   85.35
  545   AAPL 2009-01-01   90.13
  546   AAPL 2009-02-01   89.31
  547   AAPL 2009-03-01  105.12
  548   AAPL 2009-04-01  125.83
  549   AAPL 2009-05-01  135.81
  550   AAPL 2009-06-01  142.43
  551   AAPL 2009-07-01  163.39
  552   AAPL 2009-08-01  168.21
  553   AAPL 2009-09-01  185.35
  554   AAPL 2009-10-01  188.50
  555   AAPL 2009-11-01  199.91
  556   AAPL 2009-12-01  210.73
  557   AAPL 2010-01-01  192.06
  558   AAPL 2010-02-01  204.62
  559   AAPL 2010-03-01  223.02
  
  [560 rows x 3 columns],
  encoding: FacetedEncoding({
    color: Color({
      shorthand: 'symbol'
    }),
    x: X({
      shorthand: 'date'
    }),
    y: Y({
      shorthand: 'price'
    })
  }),
  mark: 'line',
  selection: {'selector002': SelectionDef({
    bind: 'scales',
    encodings: ['x', 'y'],
    type: 'interval'
  })}
})

##### Descripción del ejemplo
Un caso comun para los graficos de línea es las series de tiempo para describir algún dato en el tiempo. Para generar la serie de tiempo, solo es necesario asegurar que al eje X se le asigne una columna con fechas.

- **Línea 4**

```python
stocks = data.stocks()
```
Se guarda en la variable `stock` un DataFrame con los datos del precio de diferentes empresas durante los años.

----------------

- **Línea 9**

```python
color='symbol'
```
Utilizar el argumento `color` bajo esta marca permite  agregar multiples líneas a la visualización según el valor que tenga la columna señalada. En este caso le indicamos que le otorge un color distinto en función de la columna _symbol_.

-----------
- **Linea10**

```python
.interactive()
```
Este método transforma el gráfico estático en uno interactivo, que permite hacer _zoom_ y moverse en él. De este modo podemos observar con mayor detalle alguna zona del gráfico.

#### Actividad 8: series de tiempo

Seleccione 3 paises e indique en 1 gráfico la cantidad de inmigrantes de dichos paises por años de llegada.
